File: 61.Kosaraju'sSCCAlgo.py
# Kosaraju’s Algorithm is used to find all strongly connected components in a directed graph.
# A strongly connected component (SCC) is a maximal subset of vertices such that every vertex 
# is reachable from every other vertex in that subset.

# Explanation

# Kosaraju’s algorithm works in two main phases:

# DFS & Stack: Perform a DFS on the original graph and push nodes onto a stack in 
# the order of their finishing times.

# Transpose & DFS: Reverse all edges (transpose the graph) and perform DFS 
# in the order defined by the stack. Each DFS traversal in this phase gives one SCC.

# Why it works:

# The finishing time stack ensures we always start DFS from vertices that are 
# guaranteed to reach all nodes in their SCCs in the transposed graph.

# Transposing the graph ensures that the DFS in phase 2 only explores vertices within the same SCC.

graph = {
    0: [1],
    1: [2],
    2: [0, 3],
    3: [4],
    4: []
}

def dfs(v, visited, stack, g):
    visited[v] = True
    for u in g[v]:
        if not visited[u]:
            dfs(u, visited, stack, g)
    stack.append(v)

def transpose_graph(g):
    gt = {i : [] for i in g}
    for v in g:
        for u in g[v]:
            gt[u].append(v)
    return gt

def dfs_transposed(v, visited, gt, component):
    visited[v] = True
    component.append(v)
    for u in gt[v]:
        if not visited[u]:
            dfs_transposed(u, visited, gt, component)

def kosaraju_scc(g):
    visited = {i : False for i in g}
    stack = []

    for v in g:
        if not visited[v]:
            dfs(v, visited, stack, g)

    gt = transpose_graph(g)

    visited = {i : False for i in g}
    sccs = []

    while stack:
        v = stack.pop()
        if not visited[v]:
            component = []
            dfs_transposed(v, visited, gt, component)
            sccs.append(component)

    return sccs

result = kosaraju_scc(graph)
print("Strongly Connected Components:", result)

File: 61.Kosaraju'sSCCAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.util.*;

public class KosarajuSCC {

    // Step 1: DFS and fill stack with finishing times
    static void dfs(int v, boolean[] visited, Stack<Integer> stack, List<List<Integer>> g) {
        visited[v] = true;  // Mark current node as visited
        for (int u : g.get(v)) { // Visit all neighbors
            if (!visited[u]) {
                dfs(u, visited, stack, g);
            }
        }
        stack.push(v);  // Push vertex to stack after visiting all neighbors
    }

    // Step 2: Transpose the graph
    static List<List<Integer>> transposeGraph(List<List<Integer>> g) {
        int n = g.size();
        List<List<Integer>> gt = new ArrayList<>();
        for (int i = 0; i < n; i++) gt.add(new ArrayList<>());
        for (int v = 0; v < n; v++) {
            for (int u : g.get(v)) {
                gt.get(u).add(v); // Reverse the edge
            }
        }
        return gt;
    }

    // Step 3: DFS on transposed graph to find SCC
    static void dfsTransposed(int v, boolean[] visited, List<List<Integer>> gt, List<Integer> component) {
        visited[v] = true;
        component.add(v);
        for (int u : gt.get(v)) {
            if (!visited[u]) {
                dfsTransposed(u, visited, gt, component);
            }
        }
    }

    // Kosaraju's algorithm
    static List<List<Integer>> kosarajuSCC(List<List<Integer>> g) {
        int n = g.size();
        boolean[] visited = new boolean[n];
        Stack<Integer> stack = new Stack<>();

        // Phase 1: Fill stack with vertices by finishing times
        for (int i = 0; i < n; i++) {
            if (!visited[i]) dfs(i, visited, stack, g);
        }

        // Phase 2: Transpose the graph
        List<List<Integer>> gt = transposeGraph(g);

        // Phase 3: DFS on transposed graph in stack order
        visited = new boolean[n];
        List<List<Integer>> sccs = new ArrayList<>();

        while (!stack.isEmpty()) {
            int v = stack.pop();
            if (!visited[v]) {
                List<Integer> component = new ArrayList<>();
                dfsTransposed(v, visited, gt, component);
                sccs.add(component); // Add SCC to list
            }
        }

        return sccs;
    }

    public static void main(String[] args) {
        // Graph as adjacency list
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < 5; i++) graph.add(new ArrayList<>());
        graph.get(0).add(1);
        graph.get(1).add(2);
        graph.get(2).add(0);
        graph.get(2).add(3);
        graph.get(3).add(4);

        List<List<Integer>> result = kosarajuSCC(graph);
        System.out.println("Strongly Connected Components: " + result);
    }
}
